### PR TITLE
SI-10133 Deprecate unescaped single quote char literal

### PR DESCRIPTION
**Pull request by som-snytt**
Deprecate `'''` syntax for single quote. Error if `-Xsource:2.13`.

Also try to suggest using double quotes for `'string'` but that's quite limited.

Let `ReplReporter` handle indentation of error text (instead of indenting source); that also handles syntax errors.

### CODE REVIEW

**Comment by adriaanm** [comment]
(just reducing jenkins load until this is reworked)

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala [hunk 1]
```diff
@@ -543,24 +543,40 @@ trait Scanners extends ScannersCommon {
...
             else if (isOperatorPart(ch) && (ch != '\\'))
               charLitOr(getOperatorRest)
             else if (!isAtEnd && (ch != SU && ch != CR && ch != LF || isUnicodeEscape)) {
+              val isEmptyCharLit = (ch == '\'')
               getLitChar()
               if (ch == '\'') {
-                nextChar()
-                token = CHARLIT
-                setStrVal()
+                if (isEmptyCharLit && settings.isScala213)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: Sadly, the `isScala213` check is expensive (`def isScala213: Boolean = source.value >= version213`). This keeps coming up, maybe it's time to do something about this (/cc performance team @retronym, @lrytz).

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: Luckily it's only invoked on two successive single quotes.

By expensive, you mean more than a boolean? A couple of calls, instanceof, int compares?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: Fair point. I missed that it was in a rare error condition. We're talking lots of object allocation, regex matching, other kinds of matching and tons of method calls. I guess we could start by not creating a new ScalaVersion object every time from a string.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: I don't think it does that any more.

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: Ah, I spent my week in 2.11.x land, maybe that's why.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: Oh, is that where you were. I hope you found someone to water the plants while you were away.

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:564]: I prefer succulents anyway.

**Comment by SethTisue** [comment]
needs rebase

**Comment by som-snytt** [comment]
Letting Jenkins tell me what to do, it's an annoying rebase. Le rebase ennuyeux.

**Comment by lrytz** [comment]
Needs some `.check` file updates (hopefully that's all)

**Comment by som-snytt** [comment]
There was a clever change for `SessionTest` to source its transcript from a `.check` file, like pasting a transcript into REPL.

**Review by retronym — APPROVED** [review]

**Comment by som-snytt** [comment]
@adriaanm sorry, did that a51b9865f5bdc60dccf8ee36d2a8bf8126d48adc collide with adriaanite? I lost track.

**Comment by adriaanm** [comment]
Yes, I think I managed to incorporate its improvements into 2.13.0-M2

**Pull request merged**